plenum/server/consensus/consensus_data_provider.py

```diff
@@ -75,70 +75,72 @@ def stable_checkpoint(self) -> int:
         return self._stable_checkpoint
 
     @property
-    def checkpoints(self) -> List[Checkpoint]:
+    def checkpoints(self) -> SortedListWithKey:
         """
         List of Checkpoints, which has been sent, but not stabilized yet
         """
-        return self._checkpoints.values()
+        return self._checkpoints
 
     def preprepare_batch(self, pp: PrePrepare):
         """
         After pp had validated, it placed into _preprepared list
         """
+        # TODO: move logic checks upper the functions stack
+        if pp in self._preprepared:
+            raise LogicError('New pp cannot be stored in preprepared')
+        if pp in self._prepared:
+            raise LogicError('New pp cannot be stored in prepared')
+        if self._checkpoints and pp.ppSeqNo < self._checkpoints[-1].seqNoEnd:
+            raise LogicError('ppSeqNo cannot be lower than last checkpoint')
```

Comment:
I think it's better to make helper property `last_checkpoint` and use it here. Also I'm not sure whether we should raise logic error in the long run - probably discarding batch and logging warning would suffice, although for debugging fail fast is certainly better.

Reply:
field added